source/adapters/opencl/command_buffer.cpp

```diff
@@ -359,65 +425,213 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
+      pUpdateKernelLaunch->pNewPointerArgList;
+  std::vector<cl_mutable_dispatch_arg_khr> CLUSMArgs(NumPointerArgs);
+  for (uint32_t i = 0; i < NumPointerArgs; i++) {
+    const ur_exp_command_buffer_update_pointer_arg_desc_t &URPointerArg =
+        ArgPointerList[i];
+    cl_mutable_dispatch_arg_khr &USMArg = CLUSMArgs[i];
+    USMArg.arg_index = URPointerArg.argIndex;
+    USMArg.arg_value = *(void *const *)URPointerArg.pNewPointerArg;
   }
 
-  cl_uint ref_count;
-  CL_RETURN_ON_FAILURE(clGetCommandBufferInfoKHR(
-      hCommandBuffer->CLCommandBuffer, CL_COMMAND_BUFFER_REFERENCE_COUNT_KHR,
-      sizeof(ref_count), &ref_count, nullptr));
+  // Find the memory object and scalar arguments to the kernel.
+  const uint32_t NumMemobjArgs = pUpdateKernelLaunch->numNewMemObjArgs;
+  const ur_exp_command_buffer_update_memobj_arg_desc_t *ArgMemobjList =
+      pUpdateKernelLaunch->pNewMemObjArgList;
+  const uint32_t NumValueArgs = pUpdateKernelLaunch->numNewValueArgs;
+  const ur_exp_command_buffer_update_value_arg_desc_t *ArgValueList =
+      pUpdateKernelLaunch->pNewValueArgList;
+
+  std::vector<cl_mutable_dispatch_arg_khr> CLArgs;
+  for (uint32_t i = 0; i < NumMemobjArgs; i++) {
+    const ur_exp_command_buffer_update_memobj_arg_desc_t &URMemObjArg =
+        ArgMemobjList[i];
+    cl_mutable_dispatch_arg_khr CLArg{
+        URMemObjArg.argIndex, // arg_index
+        sizeof(cl_mem),       // arg_size
+        cl_adapter::cast<const cl_mem *>(
+            &URMemObjArg.hNewMemObjArg) // arg_value
+    };
+
+    CLArgs.push_back(CLArg);
+  }
+
+  for (uint32_t i = 0; i < NumValueArgs; i++) {
+    const ur_exp_command_buffer_update_value_arg_desc_t &URValueArg =
+        ArgValueList[i];
+    cl_mutable_dispatch_arg_khr CLArg{
+        URValueArg.argIndex,    // arg_index
+        URValueArg.argSize,     // arg_size
+        URValueArg.pNewValueArg // arg_value
+    };
+    CLArgs.push_back(CLArg);
+  }
+
+  const cl_uint NewWorkDim = pUpdateKernelLaunch->newWorkDim;
+  cl_uint &CLWorkDim = hCommand->WorkDim;
+  if (NewWorkDim != 0 && NewWorkDim != CLWorkDim) {
+    // Limitation of the cl_khr_command_buffer_mutable_dispatch specification
+    // that it is an error to change the ND-Range size.
+    // https://github.com/KhronosGroup/OpenCL-Docs/issues/1057
+    return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  }
 
-  if (pPropValue) {
-    if (propSize != sizeof(cl_uint)) {
-      return UR_RESULT_ERROR_INVALID_SIZE;
+  // Update the ND-Range configuration of the kernel.
+  const size_t CopySize = sizeof(size_t) * CLWorkDim;
+  std::vector<size_t> CLGlobalWorkOffset, CLGlobalWorkSize, CLLocalWorkSize;
+
+  if (auto GlobalWorkOffsetPtr = pUpdateKernelLaunch->pNewGlobalWorkOffset) {
+    CLGlobalWorkOffset.resize(CLWorkDim);
+    std::memcpy(CLGlobalWorkOffset.data(), GlobalWorkOffsetPtr, CopySize);
+    if (CLWorkDim < 3) {
+      const size_t ZeroSize = sizeof(size_t) * (3 - CLWorkDim);
```

Comment:
Couple instances here and below of some magic numbers, although their values might be more obvious to someone more familiar with workgroup sizes.